homeassistant/components/zwave_js/climate.py

```diff
@@ -0,0 +1,304 @@
+
+    def __init__(self, client: ZwaveClient, info: ZwaveDiscoveryInfo) -> None:
+        """Initialize lock."""
+        super().__init__(client, info)
+        self._hvac_modes: Dict[str, Optional[int]] = {}
+        self._hvac_presets: Dict[str, Optional[int]] = {}
+
+        self._current_mode = self.get_zwave_value(
+            "mode",
+            command_class=CommandClass.THERMOSTAT_MODE,
+            add_to_watched_value_ids=True,
+        )
+        self._setpoint_values: Dict[ThermostatSetpointType, ZwaveValue] = {}
+        for enum in ThermostatSetpointType:
+            self._setpoint_values[enum] = self.get_zwave_value(
+                THERMOSTAT_SETPOINT_PROPERTY,
+                command_class=CommandClass.THERMOSTAT_SETPOINT,
+                value_property_key_name=enum.value,
+                add_to_watched_value_ids=True,
+            )
+        self._operating_state = self.get_zwave_value(
+            THERMOSTAT_OPERATING_STATE_PROPERTY,
+            command_class=CommandClass.THERMOSTAT_OPERATING_STATE,
+            add_to_watched_value_ids=True,
+        )
+        self._current_temp = self.get_zwave_value(
+            THERMOSTAT_CURRENT_TEMP_PROPERTY,
+            command_class=CommandClass.SENSOR_MULTILEVEL,
+            add_to_watched_value_ids=True,
+        )
+        self._set_modes_and_presets()
+
+    def _setpoint_value(self, setpoint_type: ThermostatSetpointType) -> ZwaveValue:
+        """Optionally return a ZwaveValue for a setpoint."""
+        val = self._setpoint_values[setpoint_type]
+        if val is None:
+            raise TypeError("Value requested is not available")
+
+        return val
+
+    def _set_modes_and_presets(self) -> None:
+        """Convert Z-Wave Thermostat modes into Home Assistant modes and presets."""
+        all_modes: Dict[str, Optional[int]] = {}
+        all_presets: Dict[str, Optional[int]] = {PRESET_NONE: None}
+
+        # Z-Wave uses one list for both modes and presets.
+        # Iterate over all Z-Wave ThermostatModes and extract the hvac modes and presets.
+        current_mode = self._current_mode
+        if not current_mode:
+            return
+        for mode_id, mode_name in current_mode.metadata.states.items():
+            mode_id = int(mode_id)
+            if mode_id in THERMOSTAT_MODES:
+                # treat value as hvac mode
+                hass_mode = ZW_HVAC_MODE_MAP.get(mode_id)
+                if hass_mode:
+                    all_modes[hass_mode] = mode_id
+            else:
+                # treat value as hvac preset
+                all_presets[mode_name] = mode_id
+        self._hvac_modes = all_modes
+        self._hvac_presets = all_presets
+
+    @property
+    def _current_mode_setpoint_enums(self) -> List[ThermostatSetpointType]:
+        """Return the list of enums that are relevant to the current thermostat mode."""
+        return THERMOSTAT_MODE_SETPOINT_MAP.get(int(self._current_mode.value), [])  # type: ignore
```

Comment:
`self._current_mode` can be `None` it seems. We need to handle that. Is that why we ignore type?

Reply:
The original intent was to deal with this error: `Returning Any from function declared to return "List[Any]"  [no-any-return]`. `self._current_mode` can't practically speaking be `None` because it's the primary value. The reason I moved from using `primary_value` for this is because based on reading through the OZW integration, I'm not sure if it's the best property for discovery, so this makes it easier to switch the primary value with minimal changes

Reply:
If current mode *is* the primary value we should use that attribute.

What is the risk with using it for discovery? Are there devices that doesn't support this value?

Should we have different classes with different primary value (if needed) to make the logic more predictable?

Reply:
My understanding of the spec is that the device classes I added support for require this command class and value, so I suppose it is ok. The risk is really just from parsing the OZW integration which implies that there are thermostats that don't support this value. I will keep researching and if necessary I can open a new PR to address this, potentially expanding the device classes we can support

Reply:
Regarding the typing I think we may need to enhance the typing of the enum in the library.

Reply:
Where should the typing be enhanced? I tried some changes to the typing locally but still keep hitting the mypy error

Reply:
We can ignore it for now. I think we may need to explicitly type the Enums in the library. I've looked around a bit for best practices how to type Enums, but it's not totally clear to me yet.